util/src/overlaydb.rs

```diff
@@ -40,340 +47,178 @@ use kvdb::{Database, DBTransaction};
+		Self::new(Database::open_default(dir.to_str().unwrap()).unwrap(), mode)
 	}
 
-	/// Commit all operations to given batch.
+	/// Commit all operations to given batch. Returns the number of insertions
+	/// and deletions.
 	pub fn commit_to_batch(&mut self, batch: &DBTransaction) -> Result<u32, UtilError> {
 		let mut ret = 0u32;
 		let mut deletes = 0usize;
-		for i in self.overlay.drain().into_iter() {
-			let (key, (value, rc)) = i;
-			if rc != 0 {
-				match self.payload(&key) {
-					Some(x) => {
-						let (back_value, back_rc) = x;
-						let total_rc: i32 = back_rc as i32 + rc;
-						if total_rc < 0 {
-							return Err(From::from(BaseDataError::NegativelyReferencedHash(key)));
-						}
-						deletes += if self.put_payload_in_batch(batch, &key, (back_value, total_rc as u32)) {1} else {0};
-					}
-					None => {
-						if rc < 0 {
-							return Err(From::from(BaseDataError::NegativelyReferencedHash(key)));
-						}
-						self.put_payload_in_batch(batch, &key, (value, rc as u32));
+		for (key, (value, rc)) in self.overlay.drain() {
+			match rc {
+				0 => continue,
+				1 => try!(batch.put(&key, &value)),
+				-1 => {
+					deletes += 1;
+					if self.mode == DeletionMode::Delete {
+						try!(batch.delete(&key));
 					}
-				};
-				ret += 1;
+				}
+				rc => return Err(BaseDataError::InvalidReferenceCount(key, rc).into()),
 			}
+
+			ret += 1;
 		}
-		trace!("OverlayDB::commit() deleted {} nodes", deletes);
+
+		trace!(target: "overlaydb", "OverlayDB::commit() deleted {} nodes", deletes);
 		Ok(ret)
 	}
 
-	/// Commit all memory operations to the backing database.
-	///
-	/// Returns either an error or the number of items changed in the backing database.
-	///
-	/// Will return an error if the number of `remove()`s ever exceeds the number of
-	/// `insert()`s for any key. This will leave the database in an undeterminate
-	/// state. Don't ever let it happen.
-	///
-	/// # Example
-	/// ```
-	/// extern crate ethcore_util;
-	/// use ethcore_util::hashdb::*;
-	/// use ethcore_util::overlaydb::*;
-	/// fn main() {
-	///   let mut m = OverlayDB::new_temp();
-	///   let key = m.insert(b"foo");			// insert item.
-	///   assert!(m.contains(&key));			// key exists (in memory).
-	///   assert_eq!(m.commit().unwrap(), 1);	// 1 item changed.
-	///   assert!(m.contains(&key));			// key still exists (in backing).
-	///   m.remove(&key);							// delete item.
-	///   assert!(!m.contains(&key));			// key "doesn't exist" (though still does in backing).
-	///   m.remove(&key);							// oh dear... more removes than inserts for the key...
-	///   //m.commit().unwrap();				// this commit/unwrap would cause a panic.
-	///   m.revert();							// revert both removes.
-	///   assert!(m.contains(&key));			// key now still exists.
-	/// }
-	/// ```
+	/// Commit all operations to the backing database. Returns the number of insertions and deletions.
+	/// This does not commit auxiliary data.
```

Comment:
the reason for this is that different databases deal with auxiliary data in different ways.